filter intervals still crashing

from @zaczap
```
2018-02-21 15:30:31 Hail: INFO: interval filter loaded 89 of 1274 partitions
Traceback (most recent call last):
  File "/tmp/85fbcd66-b973-4c5c-9216-ef00b9a7f3a7/temp_vep.py", line 34, in <module>
    vds = hl.filter_intervals(vds, intervals)
  File "<decorator-gen-718>", line 2, in filter_intervals
  File "/tmp/85fbcd66-b973-4c5c-9216-ef00b9a7f3a7/hail-devel-5b95158ed055.zip/hail/utils/java.py", line 198, in handle_py4j
hail.utils.java.FatalError: requirement failed

Java stack trace:
scala.Predef$.require(Predef.scala:212)
     at is.hail.rvd.OrderedRVDPartitioner.<init>(OrderedRVDPartitioner.scala:28)
     at is.hail.rvd.OrderedRVDPartitioner.copy(OrderedRVDPartitioner.scala:98)
     at is.hail.rvd.OrderedRVD.filterIntervals(OrderedRVD.scala:270)
     at is.hail.methods.FilterIntervals$.apply(FilterIntervals.scala:23)
     at is.hail.methods.FilterIntervals$.apply(FilterIntervals.scala:18)
     at is.hail.methods.FilterIntervals.apply(FilterIntervals.scala)
     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     at java.lang.reflect.Method.invoke(Method.java:498)
     at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:237)
     at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
     at py4j.Gateway.invoke(Gateway.java:280)
     at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
     at py4j.commands.CallCommand.execute(CallCommand.java:79)
     at py4j.GatewayConnection.run(GatewayConnection.java:214)
     at java.lang.Thread.run(Thread.java:748)
Hail version: devel-5b95158
Error summary: requirement failed
ERROR: (gcloud.dataproc.jobs.submit.pyspark) Job [85fbcd66-b973-4c5c-9216-ef00b9a7f3a7] entered state [ERROR] while waiting for [DONE].
```

--- a/python/hail/methods/tests.py
+++ b/python/hail/methods/tests.py
@@ -591,13 +591,18 @@ class Tests(unittest.TestCase):
 
 
     def test_filter_intervals(self):
-        ds = self.get_dataset()
+        ds = hl.import_vcf(test_file('sample.vcf'), min_partitions=20)
+        print(ds.num_partitions())
         self.assertEqual(
             hl.filter_intervals(ds, hl.Interval.parse('20:10639222-10644705')).count_rows(), 3)
 
-        self.assertEqual(
-            hl.filter_intervals(ds, [hl.Interval.parse('20:10639222-10644700'),
-                                     hl.Interval.parse('20:10644700-10644705')]).count_rows(), 3)
+        intervals = [hl.Interval.parse('20:10639222-10644700'),
+                     hl.Interval.parse('20:10644700-10644705')]
+        self.assertEqual(hl.filter_intervals(ds, intervals).count_rows(), 3)
+
+        intervals = [hl.Interval.parse('[20:10019093-10026348]'),
+                     hl.Interval.parse('[20:17705793-17716416]')]
+        self.assertEqual(hl.filter_intervals(ds, intervals).count_rows(), 4)
 
     def test_balding_nichols_model(self):
         from hail.stats import TruncatedBetaDist

--- a/src/main/scala/is/hail/rvd/OrderedRVD.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVD.scala
@@ -265,7 +265,8 @@ class OrderedRVD private(
     if (newPartitionIndices.isEmpty)
       OrderedRVD.empty(sparkContext, typ)
     else {
-      subsetPartitions(newPartitionIndices).copy(rdd = rdd.filter(pred))
+      val sub = subsetPartitions(newPartitionIndices)
+      sub.copy(rdd = sub.rdd.filter(pred))
     }
   }
 

--- a/src/main/scala/is/hail/rvd/OrderedRVD.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVD.scala
@@ -258,18 +258,14 @@ class OrderedRVD private(
       }
     }
       .toSet // distinct
-      .toArray.sorted[Int]
+      .toArray[Int].sorted[Int]
 
     info(s"interval filter loaded ${ newPartitionIndices.length } of $nPartitions partitions")
 
     if (newPartitionIndices.isEmpty)
       OrderedRVD.empty(sparkContext, typ)
     else {
-      val newRDD = new AdjustedPartitionsRDD(rdd, newPartitionIndices.map(i => Array(Adjustment(i, (it: Iterator[RegionValue]) => it.filter(pred)))))
-      OrderedRVD(typ,
-        partitioner.copy(numPartitions = newPartitionIndices.length,
-          rangeBounds = UnsafeIndexedSeq(partitioner.rangeBoundsType, newPartitionIndices.map(partitioner.rangeBounds))),
-        newRDD)
+      subsetPartitions(newPartitionIndices).copy(rdd = rdd.filter(pred))
     }
   }
 

--- a/src/main/scala/is/hail/rvd/OrderedRVD.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVD.scala
@@ -258,7 +258,7 @@ class OrderedRVD private(
       }
     }
       .toSet // distinct
-      .toArray[Int].sorted[Int]
+      .toArray.sorted[Int]
 
     info(s"interval filter loaded ${ newPartitionIndices.length } of $nPartitions partitions")
 

--- a/src/test/scala/is/hail/variant/IntervalSuite.scala
+++ b/src/test/scala/is/hail/variant/IntervalSuite.scala
@@ -5,8 +5,10 @@ import is.hail.annotations.Annotation
 import is.hail.check.{Gen, Prop}
 import is.hail.expr.types._
 import is.hail.io.annotators.IntervalList
+import is.hail.methods.FilterIntervals
 import is.hail.utils._
 import is.hail.testUtils._
+import org.apache.spark.sql.Row
 import org.testng.annotations.Test
 
 class IntervalSuite extends SparkSuite {
@@ -299,5 +301,13 @@ class IntervalSuite extends SparkSuite {
       val actual = i0.queryOverlappingValues(ord, interval)
       assert(actual.sameElements(vals), s"\n for interval $interval, expected [${ vals.mkString(", ") }], got [${ actual.mkString(", ") }]")
     }
+  }
+
+  @Test def testFilterIntervals() {
+    val ds = hc.importVCF("src/test/resources/sample.vcf", nPartitions=Some(20))
+    val intervals = Array(Interval(Row(Locus("20", 10019093)), Row(Locus("20", 10026348)), true, true),
+      Interval(Row(Locus("20", 17705793)), Row(Locus("20", 17716416)), true, true))
+    val iTree = IntervalTree(ds.rvd.typ.pkType.ordering, intervals)
+    assert(FilterIntervals(ds, iTree, true).countVariants() == 4)
   }
 }

